raft/src/main/java/org/apache/kafka/raft/internals/BatchAccumulator.java

```diff
@@ -211,6 +214,33 @@ public long timeUntilDrain(long currentTimeMs) {
         }
     }
 
+    /**
+     * Check if the current batch size has exceeded the min flush size.
+     *
+     * Note that this method works on best effort i.e it tries to acquire the append lock and if it can't
+     * then instead of blocking, it returns false.
+     *
+     * This means that if the thread responsible for appending is holding the lock and the linger time hasn't expired
+     * yet, then even though the batch size exceeds the min flush size, the records won't be drained as the lock
+     * couldn't be acquired. This also means that in subsequent run(s), this method should be able to acquire the lock
+     * and return true in the event the linger time hasn't expired yet.
+     *
+     * @return true if the append lock could be acquired and the accumulated bytes are greater than configured min flush
+     * bytes size, false otherwise.
+     */
+    public boolean batchSizeExceedsMinFlushSize() {
```

Comment:
I think we need to look at completed batches in here as well, right? To simplify the implementation, I think we could do something like the following:

1. Inside `append`, while we are already holding the lock, we can check if the accumulated bytes (including `completed` and `currentBatch`) have reached `minFlushSize`. If so, we can call `completeCurrentBatch` to ensure that `completed` holds all the data that needs to be drained.
2. Inside `timeUntilDrain`, if the linger timer hasn't been reached, we can iterate `completed` and check if there are enough bytes to flush. Then we don't need to acquire the lock unless we need to drain.

Would that work?

Reply:
Yeah completed batches should be considered. Also, with what you have suggested, it removes the need to hold the lock again in batchSizeExceedsMinFlushSize. Basically, we won't even need this method anymore.

Reply:
> 1. Inside append, while we are already holding the lock, we can check if the accumulated bytes (including completed and currentBatch) have reached minFlushSize. If so, we can call completeCurrentBatch to ensure that completed holds all the data that needs to be drained.
> 2. Inside timeUntilDrain, if the linger timer hasn't been reached, we can iterate completed and check if there are enough bytes to flush. Then we don't need to acquire the lock unless we need to drain.

I like this suggestion.